docs/docs/typelevel.md

```diff
@@ -0,0 +1,519 @@
+   - A binding `def x = E` for a by-name parameter is dropped if `x` is not used in the rest of the expansion.
+   - A binding `def x = E` for a by-name parameter that is used exactly once in the result of the
+     expansion is dropped and the reference to `x` is replaced by `E`.
+
+   Dropping a binding might make other bindings redundant. Garbage collection proceeds until no further bindings
+   can be dropped.
+
+## Restrictions for Transparent and Typelevel Functions
+
+Transparent methods are effectively final; they may not be overwritten.
+
+If a transparent
+method has a toplevel match expression or a toplevel splice `~` on its right-hand side,
+it is classified as a typelevel method that can _only_ be executed at compile time. For typelevel methods two more restrictions apply:
+
+ 1. They must be always fully applied.
+ 2. They may override other methods only if one of the overridden methods is concrete.
+
+The right hand side of a typelevel method is never invoked by dynamic dispatch. As an example consider a situation like the following:
+```scala
+class Iterable[T] {
+  def foreach(f: T => Unit): Unit = ...
+}
+class List[T] extends Iterable[T] {
+  override transparent def foreach(f: T => Unit): Unit = ...
+}
+val xs: Iterable[T] = ...
+val ys: List[T] = ...
+val zs: Iterable[T] = ys
+xs.foreach(f)  // calls Iterable's foreach
+ys.foreach(f)  // expands to the body of List's foreach
+zs.foreach(f)  // calls Iterable's foreach
+```
+It follows that an overriding typelevel method should implement exactly the same semantics as the
+method it overrides (but possibly more effeciiently).
+
+## Matching on Types
+
+We have seen so far transparent functions that take terms (tuples and integers) as parameters. What if we want to base case distinctions on types instead? For instance, one would like to be able to write a function `defaultValue`, that, given a type `T`
+returns optionally the default value of `T`, if it exists. In fact, we can already express
+this using ordinary match expressions and a simple helper function, `scala.typelevel.erasedValue`, which is defined as follows:
+```scala
+erased def erasedValue[T]: T = ???
+```
+The `erasedValue` function _pretends_ to return a value of its type argument `T`. In fact, it would always raise a `NotImplementedError` exception when called. But the function can in fact never be called, since it is declared `erased`, so can be only used a compile-time during type checking.
+
+Using `erasedValue`, we can then define `defaultValue` as follows:
+```scala
+transparent def defaultValue[T] = erasedValue[T] match {
+  case _: Byte => Some(0: Byte)
+  case _: Char => Some(0: Char)
+  case _: Short => Some(0: Short)
+  case _: Int => Some(0)
+  case _: Long => Some(0L)
+  case _: Float => Some(0.0f)
+  case _: Double => Some(0.0d)
+  case _: Boolean => Some(false)
+  case _: Unit => Some(())
+  case _: t >: Null => Some(null)
+  case _ => None
+}
+```
```

Comment:
I think it would be much simpler to allow pattern matching on types, just like :

```
transparent def defaultValue[T] = T match {
  case Byte => Some(0: Byte)
  ...
}
```

When defaultValue called is called at compile time, it might be called before type erasure, so I think it might be possible to allow this for typelevel transparent methods.

Reply:
Ok, thanks for the explanation!